include/rmm/mr/device/sam_headroom_resource_adaptor.hpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/cuda_device.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/system_memory_resource.hpp>
+#include <rmm/resource_ref.hpp>
+
+namespace rmm::mr {
+/**
+ * @addtogroup device_resource_adaptors
+ * @{
+ * @file
+ */
+/**
+ * @brief Resource that adapts system memory resource to allocate memory with a headroom.
+ *
+ * When GPU memory is over-subscribed, system allocated memory (SAM) will migrate to the GPU
+ * and cannot migrate back, thus causing other CUDA calls to fail with out-of-memory errors. To
+ * work around this problem, we can reserve some GPU memory as headroom for other CUDA calls.
+ * Doing this check can be expensive, so only large buffers above the given threshold will be
+ * checked.
+ *
+ * @tparam Upstream Type of the upstream resource used for allocation/deallocation. Must be
+ *                  `system_memory_resource`.
+ */
+template <typename Upstream>
+class sam_headroom_resource_adaptor final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a headroom adaptor using `upstream` to satisfy allocation requests.
+   *
+   * @param upstream The resource used for allocating/deallocating device memory. Must be
+   *                 `system_memory_resource`.
+   * @param headroom_size Size of the reserved GPU memory as headroom
+   * @param threshold_size Size of the allocation above which to check for headroom
```

Comment:
Perhaps note (maybe in the class docstring):

> If `threshold_size` is non-zero, then an application making many small allocations can eat into the headroom, and run out of memory for other CUDA calls, even when using this resource adaptor.

?

Reply:
Done.